storybookjs/storybook
# Angular: Add random attribute to bootstrapped selector

Closes #23875






## What I did



Each story needs a unique selector to bootstrap the Angular app. #15501 had appended a unique id to the `storyId`, but when the `prepareForInline` function was removed, that part of it must not have been moved to a new place. I went with a similar idea, but I did it a little different. Using the same `nanoid` dependency, I added an attribute that is set to the `storyId` with a unique string appended. This attribute is then added to the element that will be bootstrapped as an attribute. The boostrap selector changes from `storyId` to `storyId[uniqueAttribute]`.

I did this in the `AbstractRenderer`, so it applies to all stories, but I think it is only needed by the `DocsRenderer`. If we don't want the `CanvasRenderer` to use the attribute in it's selector then I can move that attribute initialization to the `DocsRenderer`.

I tried to preserve the `storyId`, in case something is using it, but I didn't think about it causing two `id` DOM attributes to be the same. If that should not happen then I can try to find a spot to adjust the id, like the `prepareForInline` function used to do, but I didn't see a clear equivalent.

I also added a test for this, since there didn't seem to be one.

## Checklist for Contributors

### Testing



#### The changes in this PR are covered in the following automated tests:

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

1. Run a sandbox for template, e.g. `yarn task --task sandbox --start-from auto --template angular-cli/default-ts`
2. Add an MDX file that renders two of the same story, e.g. `<Story of={ButtonStories.Primary}/><Story of={ButtonStories.Primary}/>`
3. Open Storybook in your browser
4. Access the MDX doc and the story should be rendered twice

### Documentation



      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release



This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_
